### PR TITLE
[AAP-73866] feat: add OpenTelemetry tracing to gateway

### PR DESCRIPTION
**Pull request by chrismeyersfsu**
## Description

**Related Issue:** [AAP-73866](https://redhat.atlassian.net/browse/AAP-73866) - Enable observability in aap-gateway component

- **What is being changed?** Add OpenTelemetry distributed tracing to gateway uwsgi, grpc server, and Envoy proxy
- **Why is this change needed?** Enable end-to-end request tracing across AAP services for debugging and performance analysis
- **How does this change address the issue?** 
  - Instrument uwsgi and grpc with `setup_observability()` from django-ansible-base
  - Configure Envoy HCM with OTLP tracing to OTEL collector
  - Add observability stack sidecars (OTEL collector, Tempo, Loki, Grafana)
  - Make feature opt-in via `otel_enabled: True` config flag

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test update
- [ ] Refactoring (no functional changes)
- [x] Development environment change
- [ ] Configuration change

## Self-Review Checklist
- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [x] I have updated documentation where needed
- [x] I have considered the security impact of these changes
- [x] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [x] I have tested the changes in my local environment

### Prerequisites
- Docker or Podman
- Gateway dev environment set up

### Steps to Test
1. Uncomment `otel_enabled: True` in `tools/configs/container-startup.yml`
2. Run container startup playbook: `make up` or equivalent
3. Generate traffic to gateway (API requests, gRPC calls)
4. Open Grafana: http://localhost:3000
5. Navigate to Explore → Tempo datasource
6. Query for service traces: `{service.name="gateway-uwsgi"}` or `{service.name="gateway-grpc"}`

### Expected Results
- OTEL collector, Tempo, Loki, Grafana containers running
- Traces visible in Grafana Tempo showing gateway-uwsgi, gateway-grpc, gateway-envoy spans
- Logs correlated with trace IDs in Loki
- End-to-end request flow traceable across components

### Required Actions
- [ ] Requires downstream repository changes
  
- [ ] Requires documentation updates
  

[AAP-73866]: https://redhat.atlassian.net/browse/AAP-73866?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **New Features**
  * Observability is initialized at WSGI and gRPC startup; optional OpenTelemetry exporter support for gateways.
  * Optional monitoring stack can be enabled: OpenTelemetry Collector, Loki, Tempo, Grafana, and provisioning for them.

* **Chores**
  * Add Ansible roles, templates, defaults and Docker Compose options to provision/configure the monitoring stack.

* **Tests**
  * Improve test bootstrap for minimal environments; add tests validating observability initialization and debug startup behavior.

### CODE REVIEW

**Comment by chrismeyersfsu** [comment]
✅ **Fixed in 3f5b8f9**

Replaced hardcoded port `3100` with `{{ loki_http_port }}` variable in `loki-config.yml.j2:4`. Template now respects role defaults/overrides.

**Comment by chrismeyersfsu** [comment]
✅ **Fixed remaining port issues in a787358**

Replaced all hardcoded ports with role variables:
- **Grafana datasources.yml.j2**: `loki_http_port`, `tempo_http_port`, `otel_collector_prometheus_port`
- **Tempo config**: `tempo_http_port` for HTTP listen port
- **OTEL Collector config**: `otel_collector_grpc_port`, `otel_collector_http_port`, `otel_collector_prometheus_port`, `loki_http_port`
- **Bonus fix**: Corrected Tempo exporter endpoint format (removed `http://` prefix for gRPC)

**Comment by chrismeyersfsu** [comment]
✅ **Fixed conditional depends_on in 1cd1bc2**

Made `depends_on` blocks conditional in docker-compose template:
- **otel_collector**: only depends on loki/tempo when enabled
- **grafana**: only depends on loki/tempo when enabled

Prevents broken service references when observability components disabled.

**Comment by chrismeyersfsu** [comment]
✅ **Fixed port mappings in cb0ed90**

Container port mappings now use variables on both sides:
- **otel_collector**: `{{ otel_collector_grpc_port }}:{{ otel_collector_grpc_port }}`
- **otel_collector**: `{{ otel_collector_http_port }}:{{ otel_collector_http_port }}`
- **otel_collector**: `{{ otel_collector_prometheus_port }}:{{ otel_collector_prometheus_port }}`
- **loki**: `{{ loki_http_port }}:{{ loki_http_port }}`
- **tempo**: `{{ tempo_http_port }}:{{ tempo_http_port }}`

Port mappings now match container listen ports - overrides work correctly.

File: tools/ansible/roles/proxy-config/templates/envoy.yml.j2 [hunk 1]
```diff
@@ -69,8 +69,38 @@ static_resources:
...
+
+tracing:
+  http:
+    name: envoy.tracers.opentelemetry
+    typed_config:
+      "@type": type.googleapis.com/envoy.config.trace.v3.OpenTelemetryConfig
+      grpc_service:
+        envoy_grpc:
+          cluster_name: otel_collector
+      service_name: "gateway-envoy"
```
[the remark this answers is not in the document]
Comment (chrismeyersfsu) [reply tools/ansible/roles/proxy-config/templates/envoy.yml.j2]: I'd rather not make it conditional. I searched the  dev environment for any error or warning logs related to otel and found nothing. It's not creating any noise.

**Comment by chrismeyersfsu** [comment]
✅ **Fixed OTLP endpoint port in 2153804**

`OTEL_EXPORTER_OTLP_ENDPOINT` now uses `{{ otel_collector_grpc_port }}` variable:
- **Before**: `http://otel-collector:4317` (hardcoded)
- **After**: `http://otel-collector:{{ otel_collector_grpc_port }}` (respects overrides)

Gateway containers export traces to correct port when `otel_collector_grpc_port` changed.

**Comment by chrismeyersfsu** [comment]
✅ **Fixed Envoy tracing deprecation warning in f6fd444**

Migrated from deprecated Bootstrap.tracing to modern HttpConnectionManager pattern:

**Removed** (deprecated):
- Bootstrap-level `tracing:` block in `envoy.yml.j2`

**Added** (modern):
- `tracing.provider` in HttpConnectionManager (`xds_configs.py`)

**Result**:
- Eliminates deprecation warning: "Using deprecated option 'Bootstrap.tracing'"
- Functionally equivalent - same OTEL config, same service name
- Per-listener tracing config instead of global bootstrap config

**Ref**: [Envoy Bootstrap proto docs](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/bootstrap/v3/bootstrap.proto)

**Comment by chrismeyersfsu** [comment]
✅ **Fixed XDS test failures in 420ca57**

**Problem**:
4 XDS tests failing with protobuf parse error:
```
google.protobuf.json_format.ParseError: Can not find message descriptor 
by type_url: type.googleapis.com/envoy.config.trace.v3.OpenTelemetryConfig
```

**Root cause**:
Used hardcoded string for `@type` URL instead of protobuf descriptor constant.

**Fix**:
- Import `opentelemetry_pb2` in `common/envoy.py`
- Add `OPENTELEMETRY_CONFIG` constant from protobuf descriptor
- Use constant in `xds_configs.py` tracing config

**Tests fixed**:
- `test_xds_listener_discover_service_httpport_count`
- `test_xds_listener_discover_service_routes`
- `test_lds_listener_discover_service_service_type_auth_type`
- `test_xds_cluster_names`

**Review by john-westcott-iv — DISMISSED** [review]
## Code Review: [AAP-73866] feat: add OpenTelemetry tracing to gateway

**Verdict**: NEEDS_CHANGES | **Score**: 7.2/10 (Functionality 4.5, Security 9, Quality 8)

All findings below were validated against a running dev environment on this branch.

### Critical Findings

**1. XDS tracing config is unconditional in production code** (`xds_configs.py`)

`network_manager_filter()` always injects a `"tracing"` block pointing at an `otel_collector` cluster. This is gateway application code that runs in all deployments — not just dev tooling. With `otel_enabled: False` (default), no `otel_collector` cluster exists, but Envoy still instantiates the tracer.

**Validated:** Envoy logs confirm `instantiating a new tracer: envoy.tracers.opentelemetry` even with `otel_enabled=False` and no collector running.

The PR says the feature is "opt-in via `otel_enabled: True`" but the Python XDS code is always-on. This should be gated on a Django setting.

**2. `setup_observability()` called unconditionally with no error handling** (`wsgi.py`, `start_grpc_server.py`)

The upstream `setup_observability()` has no try/except. It runs at wsgi.py module import time. The failure mode is build-dependent:

- **Fresh build:** OTEL packages installed, app starts, spans drop silently to `localhost:4317` — noisy but survivable
- **Restart without rebuild:** OTEL packages not installed → `ImportError: cannot import name 'trace' from 'opentelemetry'` → `unable to load app 0 (callable not found or import error)` → both uWSGI and gRPC crash-loop

**Validated:** Complete application failure confirmed on restart without rebuild. The PR adds a hidden hard dependency on OTEL packages that breaks the opt-in contract.

Fix: gate behind a settings check, or at minimum wrap in try/except.

### Major Finding

**3. `setup_observability()` is explicitly non-idempotent** — upstream docstring warns calling it twice duplicates every exported log record. In uwsgi prefork mode (the default), the `wsgi.py` module-level call risks double initialization. Add an initialization guard or document `--lazy-apps` requirement.

### Minor Findings

4. **[Security]** Grafana configured with `GF_AUTH_ANONYMOUS_ORG_ROLE=Admin` and ports on `0.0.0.0` — bind to `127.0.0.1` or use `Viewer` role
5. **[Security]** OTEL Collector/Loki/Tempo ports all bound to `0.0.0.0` — bind to `127.0.0.1`
6. **[Quality]** No test coverage for XDS tracing config structure in `xds_configs.py`
7. **[Quality]** Ansible role defaults fully duplicated in `container_config.yml` (DRY violation)
8. **[Quality]** `test_wsgi.py` reload-based test is fragile

### Summary

The dev tooling is well-structured — Ansible roles, docker-compose guards, and `otel_enabled` flag all work correctly. The gap is that the Python application code (`xds_configs.py`, `wsgi.py`, `start_grpc_server.py`) doesn't respect that flag. Fixing the two Criticals (gate on settings) would close the opt-in gap and bring this to a clean review.

File: aap_gateway_api/tests/conftest.py [hunk 2]
```diff
@@ -4,12 +4,27 @@
...
+    sys.modules['ansible_base.observability'] = Mock()
+    sys.modules['ansible_base.observability'].setup_observability = Mock()
+
+if 'grpc.reflection' not in sys.modules:
+    sys.modules['grpc.reflection'] = Mock()
+    sys.modules['grpc.reflection'].enable_server_reflection = Mock()
+
+if 'debugpy' not in sys.modules:
+    sys.modules['debugpy'] = Mock()
+    sys.modules['debugpy'].listen = Mock()
```
Comment (tznamena) [inline aap_gateway_api/tests/conftest.py]: **Scope creep: `grpc.reflection` and `debugpy` mocks are unrelated to the observability feature**

Of the three `sys.modules` mocks added here, only `ansible_base.observability` is needed for this PR's production code changes. The other two are problematic:

**`grpc.reflection`** — This module is **never imported anywhere** in the production code — not on `devel` and not in this PR's changes to `start_grpc_server.py`. The mock exists solely to support the `@patch('grpc.reflection.enable_server_reflection')` decorators in the new tests, but those patches are themselves dead code: the `mock_reflection` parameter is received but never asserted on, and nothing in `_run_server()` calls `enable_server_reflection`. This appears to be cargo-culted from a different version of the gRPC server or a template.

**`debugpy`** — The production code already handles this with `try/except ImportError` (line 33–36 of `start_grpc_server.py`). The conftest mock is only needed because `test_run_server_debug_mode` uses `@patch('debugpy.listen')`. While arguably fine for that test, it could be scoped locally to the test rather than injected globally for all tests.

**Recommendation:**
- Remove the `grpc.reflection` block here entirely, and remove the corresponding `@patch('grpc.reflection.enable_server_reflection')` + `mock_reflection` params from both test functions in `test_start_grpc_server.py` — they are dead code.
- Consider scoping the `debugpy` mock to the specific test that needs it rather than globally in conftest.
- The `ansible_base.observability` mock should stay — it is essential for this PR.

Having these in `conftest.py` means every test session globally injects these mocks into `sys.modules`, which could mask real import issues in other tests.

**Comment by chrismeyersfsu** [comment]
> 1. XDS tracing config is unconditional in production code (xds_configs.py)

Code rabbit gave me this feedback and I responded.

https://github.com/ansible/jewel/pull/73#discussion_r3292855163

> 2. setup_observability() called unconditionally with no error handling (wsgi.py, start_grpc_server.py)

Yes. If it errors, we want to know. The dependent packages are not optional.

File: aap_gateway_api/tests/conftest.py [hunk 3]
```diff
@@ -90,7 +97,7 @@ def ensure_settings_bound_preferences():
     if not hasattr(settings, 'NOTIFICATION_RSS_FEED_URL'):
         setattr(settings, 'NOTIFICATION_RSS_FEED_URL', 'https://announcements.ansiblecloud.redhat.com/feed.atom')
     if not hasattr(settings, 'RUNTIME_FEATURE_FLAGS'):
-        setattr(settings, 'RUNTIME_FEATURE_FLAGS', True)
+        setattr(settings, 'RUNTIME_FEATURE_FLAGS', False)
```
Comment (bhavenst) [inline aap_gateway_api/tests/conftest.py]: **Important:** This changes `RUNTIME_FEATURE_FLAGS` from `True` to `False`, which is unrelated to OTEL and could silently alter behavior of existing tests that depend on feature flags being enabled. Should this be reverted, or if intentional, split into a separate commit with an explanation?

**Comment by bhavenst** [comment]
Reviewed — I agree with the two critical findings from the earlier review (unconditional XDS tracing config and unconditional `setup_observability()` with no error handling). Those should be gated on a setting to honor the opt-in contract.

I added one additional inline comment on an unrelated `RUNTIME_FEATURE_FLAGS` change in `conftest.py` that should be reverted or split out.

**Comment by chrismeyersfsu** [comment]
> unconditional XDS tracing config 

My thought was that it's best to have the code path always exercised. To this end, I made sure there were no noisy logs and such if there is no service on the other side (i.e. Tempo or Loki).

> and unconditional setup_observability() with no error handling

The AI is advocating for error handling to catch missing python dependencies. I argue that we want these kinds of errors to surface.

> Those should be gated on a setting to honor the opt-in contract.

The opt-in, currently, is controlled by setting `ANSIBLE_OBSERVABILITY_OTLP_ENDPOINT`. When set, the logs go to an OTLP compatible endpoint. When not set (the default), logs go no where.

File: aap_gateway_api/tests/conftest.py (hunk 3 above)
Comment (chrismeyersfsu) [reply aap_gateway_api/tests/conftest.py]: Sorry about that, unintended.

**Review by bhavenst — APPROVED** [review]
Feedback addressed — `RUNTIME_FEATURE_FLAGS` restored, and the `sys.modules` patch for debugpy is cleaner. LGTM.

**Comment by jay-steurer** [comment]
@tznamena  with follow up with 3iab about this PR

File: tools/ansible/roles/proxy-config/templates/envoy.yml.j2 [hunk 4]
```diff
@@ -69,6 +69,26 @@ static_resources:
...
+      endpoints:
+      - lb_endpoints:
+        - endpoint:
+            address:
+              socket_address:
+                address: {{ otel_collector_host | default('localhost') }}
+                port_value: {{ otel_collector_grpc_port | default(4317) }}
+    typed_extension_protocol_options:
+      envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
+        "@type": type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
```
Comment (bhavenst) [inline tools/ansible/roles/proxy-config/templates/envoy.yml.j2:88]: `otel_collector_host` isn't defined anywhere — falls back to `localhost`, which won't resolve to the collector container in Docker Compose. Should this default to `otel-collector` (the container hostname), or should `otel_collector_host` be set in `container_config.yml`?

The app-side OTEL works since the compose template uses `http://otel-collector:{{ otel_collector_grpc_port }}` directly, but Envoy traces won't reach the collector as-is.

**Review by bhavenst — COMMENTED** [review]
Re-review: previous feedback addressed (RUNTIME_FEATURE_FLAGS, mock scoping). One note on the Envoy otel_collector cluster host default — see inline.

File: aap_gateway_api/tests/conftest.py (hunk 2 above)
Comment (tznamena) [reply aap_gateway_api/tests/conftest.py]: This has been resolved in later commits, marking it resolved.

**Review by tznamena — APPROVED** [review]

**Comment by tznamena** [comment]
With John out today, I'm enabling auto-merge now, we got two approvals.

**Review by bhavenst — APPROVED** [review]
Re-review: otel_collector_host fix looks good — properly defined in role defaults and container_config.yml. LGTM.
